packages/store/src/stores/store-utils.js

```diff
@@ -106,3 +106,69 @@ export const provide = (mapStore, key, makeValue) => {
   return mapStore.get(key);
 };
 harden(provide);
+
+/**
+ * Helper for use cases in which the maker function is async. For two provide
+ * calls with the same key, one may be making when the other call starts and it
+ * would make again. (Then there'd be a collision when the second tries to store
+ * the key.) This prevents that race condition by immediately storing a Promise
+ * for the maker in an ephemeral store.
+ *
+ * Upon termination, the ephemeral store of pending makes will be lost. It's
+ * possible for termination to happen after the make completes and before it
+ * reaches durable storage.
+ *
+ * @template K
+ * @template V
+ * @param {MapStore<K, V>} durableStore
+ */
+export const makeAtomicProvider = durableStore => {
+  /** @type {Map<K, Promise<V>>} */
+  const pending = new Map();
```

Comment:
should this be a WeakMap? a WeakStore?

Reply:
If the provider crashes, the store will go away. If the provider doesn't crash, each key will be removed as it resolves.  I don't think there's a need for a weak collection.